services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+package com.linkedin.venice.controller;
+
+import com.linkedin.venice.controller.stats.ProtocolVersionAutoDetectionStats;
+import com.linkedin.venice.controllerapi.AdminOperationProtocolVersionControllerResponse;
+import com.linkedin.venice.controllerapi.ControllerClient;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.service.AbstractVeniceService;
+import com.linkedin.venice.utils.LatencyUtils;
+import io.tehuti.metrics.MetricsRepository;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This service is responsible for auto-detecting the admin operation protocol version for all clusters periodically.
+ * Each time ProtocolVersionDetectionTask runs, it will:
+ * 1. Get the current admin operation protocol versions from all controllers (parent + child) in the current cluster
+ * and find the smallest version - good version to use.
+ * 2. Get the admin operation protocol version from ZK.
+ * 3. If the version in ZK is different from the current version, update the version in ZK.
+ * To disable step 3, set the admin version for the cluster in ZK as -1.
+ */
+public class ProtocolVersionAutoDetectionService extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(ProtocolVersionAutoDetectionService.class);
+  private final AtomicBoolean stop = new AtomicBoolean(false);
+  private final Set<String> allClusters;
+  private final Map<String, ProtocolVersionAutoDetectionStats> clusterToStatsMap = new HashMap<>();
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceControllerMultiClusterConfig multiClusterConfigs;
+  final ScheduledExecutorService executor;
+
+  public ProtocolVersionAutoDetectionService(
+      VeniceParentHelixAdmin parentHelixAdmin,
+      VeniceControllerMultiClusterConfig multiClusterConfigs,
+      MetricsRepository metricsRepository,
+      Optional<Map<String, ProtocolVersionAutoDetectionStats>> clusterToStatsMap) {
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+    this.multiClusterConfigs = multiClusterConfigs;
+    this.allClusters = multiClusterConfigs.getClusters();
+    executor = Executors
+        .newScheduledThreadPool(multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionThreadCount());
+
+    if (clusterToStatsMap.isPresent()) {
+      // If the caller has provided a map of stats, use it - for testing purpose
+      this.clusterToStatsMap.putAll(clusterToStatsMap.get());
```

Comment:
You probably don't need this if the service's life cycle is controlled within cluster resource.

Reply:
Correct, not necessary anymore.